base/linalg/dense.jl

```diff
@@ -459,13 +459,17 @@ julia> exp(A)
 ```
 """
 exp(A::StridedMatrix{<:BlasFloat}) = exp!(copy(A))
-exp(A::StridedMatrix{<:Integer}) = exp!(float(A))
+exp(A::StridedMatrix{<:Union{Integer, Complex{<:Integer}}}) = exp!(float.(A))
 
 ## Destructive matrix exponential using algorithm from Higham, 2008,
 ## "Functions of Matrices: Theory and Computation", SIAM
 function exp!(A::StridedMatrix{T}) where T<:BlasFloat
     n = checksquare(A)
-    if ishermitian(A)
+    if T <: Real
```

Comment:
Perhaps
```julia
if T <: && issymmetric(A)
    # ...
elseif
    # ...
end
```

Reply:
The point is to avoid the check twice.  If it is real, we know it is either both Hermitian and symmetric or neither.

Reply:
But `issymmetric` will not be called unless `T <: Real`, so it is completely equivalent?

Reply:
Yes, but if it is real but not symmetric, then it also isn't Hermitian, so the condition in the `elseif` will be false by definition and doesn't need to be checked.

Reply:
Yes, but if it is real but not symmetric, then it also isn't Hermitian, so the condition in the `elseif` will be false by definition and doesn't need to be checked.

Reply:
Right, thanks!

Reply:
Why isn't it sufficient to just do:
```
if ishermitian(A)
    return full(exp(Hermitian(A)))
end
```
Why do we even need a separate real-`Symmetric` case?  Shouldn't `Hermitian` just have special methods for `T<:Real`?

(For that matter, why even have a `Symmetric` type if we only have specialized methods for real-symmetric?  Why not just use `Hermitian`?)

Reply:
Note also that for `T<:Real` and a non-symmetric matrix, this code calls both `issymmetric` and `ishermitian` .... i.e. it checks for the lack of symmetry twice.

Reply:
@stevengj I think you're referring to the code snippet posted by @fredrikekre, the actual code in this PR avoids this double-check by nesting the `if`s.

I don't know about why having a `Symmetric` is necessary (I don't know of a case where you have symmetric complex matrices), but I guess it's for semantic reasons?  But then maybe `Symmetric{T} where {T<:Real} = Hermitian{T}` could just be defined?

Reply:
You can have separate `exp(A::Hermitian{<:Real})` and `exp(A::Hermitian{<:Complex})` methods, so there need not be any performance penalty to using `Hermitian` for both cases.

I would prefer to just deprecate`Symmetric` entirely in favor of `Hermitian{<:Real}`.

Reply:
Explicitly checking `T<:Real` seems unnecessary to me, and explicit type-checks in Julia are usually a sign that we should be using dispatch.

We should be able to just do `Hermitian!(retmat)`, and it should do the right thing depending on whether `retmat` is real or complex.  See also the discussion in #17367.

Reply:
@stevengj I changed the dispatch of `exp` and the other trigonometric functions to `HermOrSym{<:Real}` in the most recent commit. So now there is no waste of resources there. :smile:

I asked on Slack, and there were actually some factorisation algorithms which worked for `Symmetric{<:Complex}`, so there seems to be some use for it. Though not here, obviously. A far more useful wrapper in my opinion would be `Antihermitian`, but this is probably not the right place to discuss it.